pkg/embeddedcluster/upgrade.go

```diff
@@ -0,0 +1,347 @@
+// operator, wait for the CRD to be up-to-date, and then apply the installation object.
+func runClusterUpgrade(ctx context.Context, k8sClient kubernetes.Interface, in *embeddedclusterv1beta1.Installation, license kotsv1beta1.License, versionLabel string) error {
+	var bin string
+
+	if in.Spec.AirGap {
+		image := in.Spec.Artifacts.AdditionalArtifacts["operator"]
+		if image == "" {
+			return fmt.Errorf("missing operator image in airgap artifacts")
+		}
+
+		b, err := pullUpgradeBinaryFromRegistry(ctx, k8sClient, image)
+		if err != nil {
+			return fmt.Errorf("pull upgrade binary from registry: %w", err)
+		}
+		bin = b
+	} else {
+		b, err := downloadUpgradeBinary(ctx, license, versionLabel)
+		if err != nil {
+			return fmt.Errorf("download upgrade binary: %w", err)
+		}
+		bin = b
+	}
+	defer os.RemoveAll(bin)
+
+	err := os.Chmod(bin, 0755)
+	if err != nil {
+		return fmt.Errorf("chmod upgrade binary: %w", err)
+	}
+
+	installationData, err := k8syaml.Marshal(in)
+	if err != nil {
+		return fmt.Errorf("marshal installation: %w", err)
+	}
+
+	log.Println("Running upgrade command...")
+
+	ctx, cancel := context.WithTimeout(ctx, clusterUpgradeTimeout)
+	defer cancel()
+
+	args := []string{"upgrade"}
+	if in.Spec.AirGap {
+		// TODO(upgrade): local-artifact-mirror-image should be included in the installation object
+		localArtifactMirrorImage, err := getLocalArtifactMirrorImage(ctx, k8sClient, in)
+		if err != nil {
+			return fmt.Errorf("get local artifact mirror image: %w", err)
+		}
+		args = append(args, "--local-artifact-mirror-image", localArtifactMirrorImage)
+	}
+	args = append(args, "--installation", "-")
+
+	cmd := exec.CommandContext(ctx, bin, args...)
+	cmd.Stdin = strings.NewReader(string(installationData))
+	pr, pw := io.Pipe()
+	defer pw.Close()
+	cmd.Stdout = pw
+	cmd.Stderr = pw
+	go func() {
+		defer pr.Close()
+		log.Println("Upgrade command output:")
+		scanner := bufio.NewScanner(pr)
+		for scanner.Scan() {
+			log.Println("  " + scanner.Text())
+		}
+	}()
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("run upgrade command: %w", err)
+	}
+
+	return nil
+}
+
+func downloadUpgradeBinary(ctx context.Context, license kotsv1beta1.License, versionLabel string) (string, error) {
+	tmpdir, err := os.MkdirTemp("", "embedded-cluster-artifact-*")
+	if err != nil {
+		return "", fmt.Errorf("create temp dir: %w", err)
+	}
+
+	req, err := newDownloadUpgradeBinaryRequest(ctx, license, versionLabel)
+	if err != nil {
+		return "", fmt.Errorf("new download upgrade binary request: %w", err)
+	}
+
+	log.Printf("Downloading upgrade binary from %s...", req.URL)
+
+	httpClient := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			if strings.Contains(req.URL.Path, "/embedded-cluster-public-files/") {
+				// Workaround as cloudflare complains about unsupported auth method
+				req.Header.Del("Authorization")
```

Comment:
it's probably good practice to always remove the auth header on redirect. idk why Go doesn't do that automatically. can probably just always delete the auth header without checking the path.

Reply:
im fixing this in the proxy https://github.com/replicatedhq/cloudflare-workers/pull/348/files